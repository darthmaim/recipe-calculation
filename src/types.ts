import { NestedRecipe } from '@gw2efficiency/recipe-nesting'

export type ExtendRecipeTree<TBase, TExtend> = Omit<TBase, 'type' | 'components'> & {
  type: 'Recipe' | 'Item' | 'Currency'
  components?: Array<ExtendRecipeTree<TBase, TExtend>>
} & TExtend

export type RecipeTree = ExtendRecipeTree<NestedRecipe, { __never: never }>

export type RecipeTreeWithQuantity = ExtendRecipeTree<
  RecipeTree,
  {
    output: number
    totalQuantity: number
    usedQuantity: number
  }
>

export type RecipeTreeWithPrices = ExtendRecipeTree<
  RecipeTreeWithQuantity,
  {
    buyPriceEach: number | false
    buyPrice: number | false
    decisionPrice: number | false
    craftPrice?: number
  }
>

export type RecipeTreeWithCraftFlags = ExtendRecipeTree<
  RecipeTreeWithPrices,
  {
    craft: boolean
  }
>

export type ItemPrices = Record<string, number>

export type AvailableItems = Record<string, number>

export type ForceBuyItems = Array<number>
