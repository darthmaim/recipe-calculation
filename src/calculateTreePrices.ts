import { RecipeTreeWithCraftFlags, RecipeTreeWithPrices, RecipeTreeWithQuantity } from './types'

// Update the tree prices
export function calculateTreePrices(
  tree: RecipeTreeWithCraftFlags,
  itemPrices: Record<string, number>
): RecipeTreeWithCraftFlags
export function calculateTreePrices(
  tree: RecipeTreeWithQuantity,
  itemPrices: Record<string, number>
): RecipeTreeWithPrices
export function calculateTreePrices(
  tree: RecipeTreeWithQuantity | RecipeTreeWithCraftFlags,
  itemPrices: Record<string, number>
): RecipeTreeWithPrices | RecipeTreeWithCraftFlags {
  // Calculate the buy prices
  let buyPriceEach = itemPrices[tree.id] || false
  if (tree.type === 'Currency') {
    // Coin is 1c per used quantity, the other currencies are not considered
    buyPriceEach = tree.id === 1 ? 1 : false
  }
  const buyPrice = buyPriceEach ? tree.usedQuantity * buyPriceEach : false
  let decisionPrice = buyPrice

  if (!tree.components) {
    return { ...tree, components: undefined, buyPriceEach, buyPrice, decisionPrice }
  }

  // Calculate the tree prices traversal for the sub-components
  const components = tree.components.map((component) => calculateTreePrices(component, itemPrices))

  // Calculate the craft price out of the best prices
  const craftPrice = components.map((c) => c.decisionPrice || 0).reduce((a, b) => a + b, 0)

  // Update the decision price of this tree segment to the craft price,
  // used to determine the craft price of the higher-up recipe
  if (
    !('craft' in tree && tree.craft === false) &&
    (('craft' in tree && tree.craft === true) || !buyPrice || craftPrice < buyPrice)
  ) {
    decisionPrice = craftPrice
  }

  return { ...tree, components, buyPriceEach, buyPrice, decisionPrice, craftPrice }
}
