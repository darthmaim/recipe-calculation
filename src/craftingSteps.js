export default function craftingStepsWrapper (tree) {
  let steps = craftingSteps(tree).reverse()

  // We don't care about steps where we have to craft "0" of something
  // This can happen when a crafting tree is updated after the initial calculation
  steps = steps.filter(step => step.quantity > 0)

  // Calculate how many times you actually have to click on "craft"
  // for items with output > 1
  steps = steps.map(step => {
    step.crafts = Math.ceil(step.quantity / step.output)
    delete step.output
    return step
  })

  // Make sure that "Mystic Clovers" get crafted as the first step,
  // since they generate items that are always useful for crafting the other steps
  let mysticCloverId = 19675
  let clovers = steps.find(step => step.id === mysticCloverId)
  if (clovers) {
    steps = steps.filter(step => step.id !== mysticCloverId)
    steps.unshift(clovers)
  }

  return steps
}

// Generate an ordered list of crafting steps
function craftingSteps (tree, steps = [], index = 0) {
  // Skip any tree parts where nothing needs to be crafted
  if (!tree.components || tree.craft === false) {
    return steps
  }

  // Go through the existing steps, and if we already have a step
  // with this id, just add up the quantities
  let stepIndex = steps.findIndex(step => step.id === tree.id)
  if (stepIndex !== -1) {
    steps[stepIndex].quantity += tree.usedQuantity
    steps[stepIndex].components = steps[stepIndex].components
      .map(component => {
        let treeComponent = tree.components.find(tC => tC.id === component.id)
        component.quantity += treeComponent.totalQuantity
        return component
      })
    index = stepIndex
  }

  // We don't have a step like this yet, push a new one at the given index
  if (stepIndex === -1) {
    steps.splice(index, 0, {
      id: tree.id,
      output: tree.output,
      quantity: tree.usedQuantity,
      minRating: tree.min_rating,
      disciplines: tree.disciplines,
      recipeId: tree.recipe_id,
      components: tree.components.map(component => {
        return {id: component.id, quantity: component.totalQuantity}
      })
    })
  }

  // Go through the components and push them after the index of their parent
  tree.components.map(component => craftingSteps(component, steps, index + 1))
  return steps
}
