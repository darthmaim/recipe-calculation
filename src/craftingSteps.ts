import { RecipeTreeWithCraftFlags } from './types'

export function craftingSteps(tree: RecipeTreeWithCraftFlags) {
  let steps = craftingStepsInner(tree).reverse()

  // We don't care about steps where we have to craft "0" of something
  // This can happen when a crafting tree is updated after the initial calculation
  steps = steps.filter((step) => step.quantity > 0)

  // Make sure that "Mystic Clovers" get crafted as the first step,
  // since they generate items that are always useful for crafting the other steps
  const mysticCloverId = 19675
  const clovers = steps.find((step) => step.id === mysticCloverId)
  if (clovers) {
    steps = steps.filter((step) => step.id !== mysticCloverId)
    steps.unshift(clovers)
  }

  return steps.map((step) => ({
    ...step,
    // Calculate how many times you actually have to click on "craft"
    // for items with output > 1 (calculate here when all steps are aggregated)
    crafts: Math.ceil(step.quantity / step.output),
    output: undefined,
  }))
}

interface CraftingStep {
  id: number
  output: number
  quantity: number
  minRating: number | null
  disciplines: Array<string>
  recipeId?: number
  components: Array<{ id: number; quantity: number }>
}

// Generate an ordered list of crafting steps
function craftingStepsInner(
  tree: RecipeTreeWithCraftFlags,
  steps: Array<CraftingStep> = [],
  index = 0
) {
  const treeComponents = tree.components

  // Skip any tree parts where nothing needs to be crafted
  if (!treeComponents || tree.craft === false) {
    return steps
  }

  // Go through the existing steps, and if we already have a step
  // with this id, just add up the quantities
  const stepIndex = steps.findIndex((step) => step.id === tree.id)
  if (stepIndex !== -1) {
    steps[stepIndex].quantity += tree.usedQuantity
    steps[stepIndex].components = steps[stepIndex].components.map((component) => {
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const treeComponent = treeComponents.find((x) => x.id === component.id)!

      component.quantity += treeComponent.totalQuantity
      return component
    })
    index = stepIndex
  }

  // We don't have a step like this yet, push a new one at the given index
  if (stepIndex === -1) {
    steps.splice(index, 0, {
      id: tree.id,
      output: tree.output,
      quantity: tree.usedQuantity,
      minRating: tree.min_rating,
      disciplines: tree.disciplines,
      recipeId: tree.recipe_id,
      components: treeComponents.map((component) => {
        return { id: component.id, quantity: component.totalQuantity }
      }),
    })
  }

  // Go through the components and push them after the index of their parent
  treeComponents.map((component) => craftingStepsInner(component, steps, index + 1))
  return steps
}
