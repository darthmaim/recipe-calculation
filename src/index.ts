export { cheapestTree } from './cheapestTree'
export { updateTree } from './updateTree'
export { usedItems } from './usedItems'
export { craftingSteps } from './craftingSteps'
export { recipeItems } from './helpers/recipeItems'
export { dailyCooldowns } from './helpers/dailyCooldowns'
export { useVendorPrices } from './helpers/useVendorPrices'
import { DAILY_COOLDOWNS } from './static/dailyCooldowns'
import { VENDOR_ITEMS } from './static/vendorItems'

export const staticItems = {
  dailyCooldowns: DAILY_COOLDOWNS.map((x) => x.id),
  buyableDailyCooldowns: DAILY_COOLDOWNS.filter((x) => x.tradable).map((x) => x.id),
  vendorItems: VENDOR_ITEMS,
}
