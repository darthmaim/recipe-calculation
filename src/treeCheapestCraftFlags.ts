import { RecipeTreeWithCraftFlags, RecipeTreeWithPrices } from './types'

// Set the craft flags based on the cheapest price
export function treeCheapestCraftFlags(
  tree: RecipeTreeWithPrices,
  forceBuyItems: Array<number>
): RecipeTreeWithCraftFlags {
  // Craft the item if it can't be bought or it is cheaper to craft
  const shouldForceBuy = forceBuyItems.indexOf(tree.id) !== -1
  const isUsed = tree.usedQuantity !== 0
  const craftIsCheaper =
    typeof tree.craftPrice !== 'undefined' && (!tree.buyPrice || tree.craftPrice < tree.buyPrice)

  const craft = !!tree.components && !shouldForceBuy && isUsed && craftIsCheaper

  if (!tree.components) {
    return { ...tree, components: undefined, craft }
  }

  // Adjust the flags for all tree's sub-recipes
  const components = tree.components.map((component) =>
    treeCheapestCraftFlags(component, forceBuyItems)
  )

  return { ...tree, components, craft }
}
