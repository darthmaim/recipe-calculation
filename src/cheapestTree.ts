import { calculateTreeQuantity } from './calculateTreeQuantity'
import { calculateTreePrices } from './calculateTreePrices'
import { calculateTreeCraftFlags } from './calculateTreeCraftFlags'
import { RecipeTree, RecipeTreeWithCraftFlags, RecipeTreeWithPrices } from './types'
import {
  NestedRecipe,
  BasicCurrencyComponent,
  BasicItemComponent,
} from '@gw2efficiency/recipe-nesting'

export function cheapestTree(
  amount: number,
  tree: NestedRecipe,
  itemPrices: Record<string, number>,
  availableItems: Record<string, number> = {},
  forceBuyItems: Array<number> = [],
  valueOwnItems: false
): RecipeTreeWithCraftFlags {
  // calculateTreeQuantity already checks for craft flags, so we can set them here when valuing owned items
  if (valueOwnItems) {
    const treeWithQuantityWithoutAvailableItems = calculateTreeQuantity(
      amount,
      tree as RecipeTree,
      {}
    )
    const treeWithPriceWithoutAvailableItems = calculateTreePrices(
      treeWithQuantityWithoutAvailableItems,
      itemPrices
    )
    const cheaperToBuyItemIds = getCheaperToBuyItemIds(treeWithPriceWithoutAvailableItems)
    disableCraftForItemIds(tree, cheaperToBuyItemIds)
  }
  // Adjust the tree total and used quantities
  const treeWithQuantity = calculateTreeQuantity(amount, tree as RecipeTree, availableItems)

  // Set the initial craft flags based on the subtree prices
  const treeWithPrices = calculateTreePrices(treeWithQuantity, itemPrices)
  const treeWithCraftFlags = calculateTreeCraftFlags(treeWithPrices, forceBuyItems)

  // Force the root to be crafted
  treeWithCraftFlags.craft = true

  // After the "craft" flags are set, update the used materials
  // to only be used for things that actually get crafted
  const treeWithQuantityPostFlags = calculateTreeQuantity(
    amount,
    treeWithCraftFlags,
    availableItems
  )

  // Recalculate the correct tree price
  return calculateTreePrices(treeWithQuantityPostFlags, itemPrices)
}

function getCheaperToBuyItemIds(
  tree: RecipeTreeWithPrices,
  ids: Array<number> = []
): Array<number> {
  // If buying the item costs less than the possible profit made by selling its craft components (with TP tax), we should buy instead of crafting
  if (
    typeof tree.craftDecisionPrice === 'number' &&
    typeof tree.buyPrice === 'number' &&
    tree.buyPrice < tree.craftDecisionPrice * 0.85
  ) {
    if (!ids.includes(tree.id)) {
      ids.push(tree.id)
    }
  }

  if (tree.components && Array.isArray(tree.components)) {
    tree.components.forEach((component) => getCheaperToBuyItemIds(component, ids))
  }

  return ids
}

type NestedRecipeAndBasicComponentWithCraftFlag = (
  | NestedRecipe
  | BasicItemComponent
  | BasicCurrencyComponent
) & {
  craft?: boolean
}

function disableCraftForItemIds(
  tree: NestedRecipeAndBasicComponentWithCraftFlag,
  ids: Array<number>
) {
  if (ids.includes(tree.id)) {
    tree.craft = false
  }

  if ('components' in tree && Array.isArray(tree.components)) {
    tree.components.forEach((component) => {
      disableCraftForItemIds(component as NestedRecipeAndBasicComponentWithCraftFlag, ids)
    })
  }
}
