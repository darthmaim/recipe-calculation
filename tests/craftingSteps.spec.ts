import { craftingSteps } from '../src'
import { RecipeTreeWithCraftFlags } from '../src/types'

describe('craftingSteps', () => {
  it('gets the correct crafting steps', () => {
    const tree: RecipeTreeWithCraftFlags = {
      id: 1,
      craft: true,
      totalQuantity: 2,
      usedQuantity: 2,
      output: 2,
      type: 'Recipe',
      quantity: 2,
      min_rating: null,
      buyPriceEach: 1,
      buyPrice: 1,
      decisionPrice: 1,
      disciplines: [],
      components: [
        {
          id: 7,
          craft: true,
          totalQuantity: 6,
          usedQuantity: 6,
          quantity: 6,
          type: 'Recipe',
          buyPriceEach: 1,
          buyPrice: 1,
          decisionPrice: 1,
          output: 1,
          min_rating: 400,
          disciplines: ['Armorsmith'],
          recipe_id: 123,
          components: [
            {
              id: 4,
              craft: false,
              totalQuantity: 12,
              usedQuantity: 12,
              type: 'Item',
              quantity: 12,
              output: 1,
              min_rating: null,
              buyPriceEach: 1,
              buyPrice: 1,
              decisionPrice: 1,
              disciplines: [],
            },
          ],
        },
        {
          id: 3,
          craft: true,
          totalQuantity: 10,
          usedQuantity: 3,
          output: 1,
          type: 'Recipe',
          quantity: 10,
          min_rating: null,
          disciplines: [],
          buyPrice: 1,
          buyPriceEach: 1,
          decisionPrice: 1,
          components: [
            {
              id: 19,
              craft: true,
              totalQuantity: 3,
              usedQuantity: 3,
              output: 1,
              type: 'Recipe',
              quantity: 3,
              min_rating: null,
              disciplines: [],
              buyPrice: 1,
              buyPriceEach: 1,
              decisionPrice: 1,
              components: [
                {
                  id: 7,
                  craft: true,
                  totalQuantity: 3,
                  usedQuantity: 3,
                  output: 1,
                  type: 'Recipe',
                  quantity: 3,
                  min_rating: null,
                  disciplines: [],
                  buyPrice: 1,
                  buyPriceEach: 1,
                  decisionPrice: 1,
                  components: [
                    {
                      id: 4,
                      craft: false,
                      totalQuantity: 6,
                      usedQuantity: 6,
                      type: 'Item',
                      quantity: 6,
                      output: 1,
                      min_rating: null,
                      buyPriceEach: 1,
                      buyPrice: 1,
                      decisionPrice: 1,
                      disciplines: [],
                    },
                  ],
                },
              ],
            },
          ],
        },
        {
          id: 5,
          craft: false,
          totalQuantity: 4,
          usedQuantity: 0,
          output: 1,
          type: 'Recipe',
          quantity: 4,
          min_rating: null,
          buyPriceEach: 1,
          buyPrice: 1,
          decisionPrice: 1,
          disciplines: [],
          components: [
            {
              id: 6,
              craft: false,
              totalQuantity: 0,
              usedQuantity: 0,
              type: 'Item',
              quantity: 0,
              output: 1,
              min_rating: null,
              buyPriceEach: 1,
              buyPrice: 1,
              decisionPrice: 1,
              disciplines: [],
            },
          ],
        },
        {
          id: 15,
          craft: true,
          totalQuantity: 1,
          usedQuantity: 1,
          output: 1,
          type: 'Recipe',
          quantity: 1,
          min_rating: null,
          buyPriceEach: 1,
          buyPrice: 1,
          decisionPrice: 1,
          disciplines: [],
          components: [
            {
              id: 7,
              craft: true,
              totalQuantity: 6,
              usedQuantity: 6,
              output: 1,
              type: 'Recipe',
              quantity: 6,
              min_rating: null,
              buyPriceEach: 1,
              buyPrice: 1,
              decisionPrice: 1,
              disciplines: [],
              components: [
                {
                  id: 4,
                  craft: false,
                  totalQuantity: 12,
                  usedQuantity: 12,
                  type: 'Item',
                  quantity: 12,
                  output: 1,
                  min_rating: null,
                  buyPriceEach: 1,
                  buyPrice: 1,
                  decisionPrice: 1,
                  disciplines: [],
                },
              ],
            },
          ],
        },
        {
          id: 1337,
          craft: true,
          totalQuantity: 1,
          usedQuantity: 0,
          output: 1,
          type: 'Recipe',
          quantity: 1,
          min_rating: null,
          buyPriceEach: 1,
          buyPrice: 1,
          decisionPrice: 1,
          disciplines: [],
          components: [
            {
              id: 42,
              craft: true,
              totalQuantity: 6,
              usedQuantity: 6,
              output: 1,
              type: 'Item',
              quantity: 6,
              min_rating: null,
              buyPriceEach: 1,
              buyPrice: 1,
              decisionPrice: 1,
              disciplines: [],
            },
          ],
        },
        {
          id: 1,
          craft: true,
          totalQuantity: 6,
          usedQuantity: 6,
          output: 1,
          type: 'Currency',
          quantity: 6,
          min_rating: null,
          buyPriceEach: 1,
          buyPrice: 1,
          decisionPrice: 1,
          disciplines: [],
        },
      ],
    }

    const usedItemObject = craftingSteps(tree)
    expect(usedItemObject).toMatchSnapshot()
  })

  it('gets the correct crafting steps with mystic clovers', () => {
    const tree: RecipeTreeWithCraftFlags = {
      id: 1,
      craft: true,
      totalQuantity: 2,
      usedQuantity: 2,
      output: 2,
      type: 'Recipe',
      quantity: 2,
      min_rating: null,
      buyPriceEach: 1,
      buyPrice: 1,
      decisionPrice: 1,
      disciplines: [],
      components: [
        {
          id: 7,
          craft: true,
          totalQuantity: 6,
          usedQuantity: 6,
          quantity: 6,
          type: 'Recipe',
          buyPriceEach: 1,
          buyPrice: 1,
          decisionPrice: 1,
          output: 1,
          min_rating: 400,
          disciplines: ['Armorsmith'],
          recipe_id: 123,
          components: [
            {
              id: 4,
              craft: false,
              totalQuantity: 12,
              usedQuantity: 12,
              type: 'Item',
              quantity: 12,
              output: 1,
              min_rating: null,
              buyPriceEach: 1,
              buyPrice: 1,
              decisionPrice: 1,
              disciplines: [],
            },
          ],
        },
        {
          id: 3,
          craft: true,
          totalQuantity: 10,
          usedQuantity: 3,
          output: 1,
          type: 'Recipe',
          quantity: 10,
          min_rating: null,
          disciplines: [],
          buyPrice: 1,
          buyPriceEach: 1,
          decisionPrice: 1,
          components: [
            {
              id: 19,
              craft: true,
              totalQuantity: 3,
              usedQuantity: 3,
              output: 1,
              type: 'Recipe',
              quantity: 3,
              min_rating: null,
              disciplines: [],
              buyPrice: 1,
              buyPriceEach: 1,
              decisionPrice: 1,
              components: [
                {
                  id: 19675,
                  craft: true,
                  totalQuantity: 3,
                  usedQuantity: 3,
                  output: 1,
                  type: 'Recipe',
                  quantity: 3,
                  min_rating: null,
                  disciplines: [],
                  buyPrice: 1,
                  buyPriceEach: 1,
                  decisionPrice: 1,
                  components: [
                    {
                      id: 19925,
                      type: 'Recipe',
                      quantity: 1,
                      output: 1,
                      components: [
                        {
                          id: 19,
                          type: 'Currency',
                          quantity: 25,
                          output: 1,
                          totalQuantity: 250,
                          usedQuantity: 250,
                          buyPriceEach: false,
                          buyPrice: false,
                          decisionPrice: false,
                          craft: false,
                          min_rating: null,
                          disciplines: [],
                        },
                        {
                          id: 2,
                          type: 'Currency',
                          quantity: 1050,
                          output: 1,
                          totalQuantity: 10500,
                          usedQuantity: 10500,
                          buyPriceEach: false,
                          buyPrice: false,
                          decisionPrice: false,
                          craft: false,
                          min_rating: null,
                          disciplines: [],
                        },
                      ],
                      min_rating: null,
                      disciplines: ['Merchant'],
                      totalQuantity: 10,
                      usedQuantity: 10,
                      buyPriceEach: false,
                      buyPrice: false,
                      decisionPrice: 0,
                      craftPrice: 0,
                      craft: true,
                    },
                    {
                      id: 19976,
                      type: 'Item',
                      quantity: 1,
                      output: 1,
                      totalQuantity: 10,
                      usedQuantity: 10,
                      buyPriceEach: 8592,
                      buyPrice: 85920,
                      decisionPrice: 85920,
                      craft: false,
                      min_rating: null,
                      disciplines: [],
                    },
                    {
                      id: 19721,
                      type: 'Item',
                      quantity: 1,
                      output: 1,
                      totalQuantity: 10,
                      usedQuantity: 10,
                      buyPriceEach: 1329,
                      buyPrice: 13290,
                      decisionPrice: 13290,
                      craft: false,
                      min_rating: null,
                      disciplines: [],
                    },
                    {
                      id: 20796,
                      type: 'Recipe',
                      quantity: 6,
                      output: 10,
                      components: [
                        {
                          id: 23,
                          type: 'Currency',
                          quantity: 1,
                          output: 1,
                          totalQuantity: 6,
                          usedQuantity: 6,
                          buyPriceEach: false,
                          buyPrice: false,
                          decisionPrice: false,
                          craft: false,
                          min_rating: null,
                          disciplines: [],
                        },
                      ],
                      min_rating: null,
                      disciplines: ['Merchant'],
                      totalQuantity: 60,
                      usedQuantity: 60,
                      buyPriceEach: false,
                      buyPrice: false,
                      decisionPrice: 0,
                      craftPrice: 0,
                      craft: true,
                    },
                  ],
                },
              ],
            },
          ],
        },
        {
          id: 5,
          craft: false,
          totalQuantity: 4,
          usedQuantity: 0,
          output: 1,
          type: 'Recipe',
          quantity: 4,
          min_rating: null,
          buyPriceEach: 1,
          buyPrice: 1,
          decisionPrice: 1,
          disciplines: [],
          components: [
            {
              id: 6,
              craft: false,
              totalQuantity: 0,
              usedQuantity: 0,
              type: 'Item',
              quantity: 0,
              output: 1,
              min_rating: null,
              buyPriceEach: 1,
              buyPrice: 1,
              decisionPrice: 1,
              disciplines: [],
            },
          ],
        },
        {
          id: 15,
          craft: true,
          totalQuantity: 1,
          usedQuantity: 1,
          output: 1,
          type: 'Recipe',
          quantity: 1,
          min_rating: null,
          buyPriceEach: 1,
          buyPrice: 1,
          decisionPrice: 1,
          disciplines: [],
          components: [
            {
              id: 7,
              craft: true,
              totalQuantity: 6,
              usedQuantity: 6,
              output: 1,
              type: 'Recipe',
              quantity: 6,
              min_rating: null,
              buyPriceEach: 1,
              buyPrice: 1,
              decisionPrice: 1,
              disciplines: [],
              components: [
                {
                  id: 4,
                  craft: false,
                  totalQuantity: 12,
                  usedQuantity: 12,
                  type: 'Item',
                  quantity: 12,
                  output: 1,
                  min_rating: null,
                  buyPriceEach: 1,
                  buyPrice: 1,
                  decisionPrice: 1,
                  disciplines: [],
                },
              ],
            },
          ],
        },
      ],
    }

    const usedItemObject = craftingSteps(tree)
    expect(usedItemObject).toMatchSnapshot()
  })

  it('gets the correct crafting steps when merging components', () => {
    const tree: RecipeTreeWithCraftFlags = {
      id: 1,
      craft: true,
      totalQuantity: 1,
      usedQuantity: 1,
      output: 1,
      type: 'Recipe',
      quantity: 1,
      min_rating: null,
      disciplines: [],
      buyPrice: 1,
      buyPriceEach: 1,
      decisionPrice: 1,
      components: [
        {
          id: 2,
          craft: true,
          totalQuantity: 1,
          usedQuantity: 1,
          output: 1,
          type: 'Recipe',
          quantity: 1,
          min_rating: null,
          disciplines: [],
          buyPrice: 1,
          buyPriceEach: 1,
          decisionPrice: 1,
          components: [
            {
              id: 3,
              craft: false,
              totalQuantity: 1,
              usedQuantity: 0,
              output: 1,
              type: 'Recipe',
              quantity: 1,
              min_rating: null,
              disciplines: [],
              buyPrice: 1,
              buyPriceEach: 1,
              decisionPrice: 1,
              components: [
                {
                  id: 4,
                  craft: false,
                  totalQuantity: 1,
                  usedQuantity: 1,
                  output: 1,
                  type: 'Item',
                  quantity: 1,
                  min_rating: null,
                  buyPriceEach: 1,
                  buyPrice: 1,
                  decisionPrice: 1,
                  disciplines: [],
                },
              ],
            },
            {
              id: 5,
              craft: true,
              totalQuantity: 1,
              usedQuantity: 1,
              output: 1,
              type: 'Recipe',
              quantity: 1,
              min_rating: null,
              disciplines: [],
              buyPrice: 1,
              buyPriceEach: 1,
              decisionPrice: 1,
              components: [
                {
                  id: 6,
                  craft: false,
                  totalQuantity: 1,
                  usedQuantity: 1,
                  output: 1,
                  type: 'Item',
                  quantity: 1,
                  min_rating: null,
                  buyPriceEach: 1,
                  buyPrice: 1,
                  decisionPrice: 1,
                  disciplines: [],
                },
              ],
            },
          ],
        },
        {
          id: 2,
          craft: true,
          totalQuantity: 1,
          usedQuantity: 1,
          output: 1,
          type: 'Recipe',
          quantity: 1,
          min_rating: null,
          disciplines: [],
          buyPrice: 1,
          buyPriceEach: 1,
          decisionPrice: 1,
          components: [
            {
              id: 3,
              craft: true,
              totalQuantity: 1,
              usedQuantity: 1,
              output: 1,
              type: 'Recipe',
              quantity: 1,
              min_rating: null,
              disciplines: [],
              buyPrice: 1,
              buyPriceEach: 1,
              decisionPrice: 1,
              components: [
                {
                  id: 4,
                  craft: false,
                  totalQuantity: 1,
                  usedQuantity: 1,
                  output: 1,
                  type: 'Item',
                  quantity: 1,
                  min_rating: null,
                  buyPriceEach: 1,
                  buyPrice: 1,
                  decisionPrice: 1,
                  disciplines: [],
                },
              ],
            },
            {
              id: 5,
              craft: true,
              totalQuantity: 1,
              usedQuantity: 1,
              output: 1,
              type: 'Recipe',
              quantity: 1,
              min_rating: null,
              disciplines: [],
              buyPrice: 1,
              buyPriceEach: 1,
              decisionPrice: 1,
              components: [
                {
                  id: 6,
                  craft: false,
                  totalQuantity: 1,
                  usedQuantity: 1,
                  output: 1,
                  type: 'Item',
                  quantity: 1,
                  min_rating: null,
                  buyPriceEach: 1,
                  buyPrice: 1,
                  decisionPrice: 1,
                  disciplines: [],
                },
              ],
            },
          ],
        },
      ],
    }

    const usedItemObject = craftingSteps(tree)

    expect(usedItemObject).toMatchSnapshot()
  })
})
