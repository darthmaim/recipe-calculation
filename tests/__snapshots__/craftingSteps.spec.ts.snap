// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`craftingSteps gets the correct crafting steps 1`] = `
Array [
  Object {
    "components": Array [
      Object {
        "id": 4,
        "quantity": 30,
        "type": "Item",
      },
    ],
    "crafts": 15,
    "disciplines": Array [
      "Armorsmith",
    ],
    "id": 7,
    "minRating": 400,
    "output": undefined,
    "quantity": 15,
    "recipeId": 123,
    "type": "Recipe",
  },
  Object {
    "components": Array [
      Object {
        "id": 7,
        "quantity": 3,
        "type": "Recipe",
      },
    ],
    "crafts": 3,
    "disciplines": Array [],
    "id": 19,
    "minRating": null,
    "output": undefined,
    "quantity": 3,
    "recipeId": undefined,
    "type": "Recipe",
  },
  Object {
    "components": Array [
      Object {
        "id": 19,
        "quantity": 3,
        "type": "Recipe",
      },
    ],
    "crafts": 3,
    "disciplines": Array [],
    "id": 3,
    "minRating": null,
    "output": undefined,
    "quantity": 3,
    "recipeId": undefined,
    "type": "Recipe",
  },
  Object {
    "components": Array [
      Object {
        "id": 7,
        "quantity": 6,
        "type": "Recipe",
      },
    ],
    "crafts": 1,
    "disciplines": Array [],
    "id": 15,
    "minRating": null,
    "output": undefined,
    "quantity": 1,
    "recipeId": undefined,
    "type": "Recipe",
  },
  Object {
    "components": Array [
      Object {
        "id": 7,
        "quantity": 6,
        "type": "Recipe",
      },
      Object {
        "id": 3,
        "quantity": 10,
        "type": "Recipe",
      },
      Object {
        "id": 5,
        "quantity": 4,
        "type": "Recipe",
      },
      Object {
        "id": 15,
        "quantity": 1,
        "type": "Recipe",
      },
      Object {
        "id": 1337,
        "quantity": 1,
        "type": "Recipe",
      },
      Object {
        "id": 1,
        "quantity": 6,
        "type": "Currency",
      },
    ],
    "crafts": 1,
    "disciplines": Array [],
    "id": 1,
    "minRating": null,
    "output": undefined,
    "quantity": 2,
    "recipeId": undefined,
    "type": "Recipe",
  },
]
`;

exports[`craftingSteps gets the correct crafting steps when merging components 1`] = `
Array [
  Object {
    "components": Array [
      Object {
        "id": 6,
        "quantity": 2,
        "type": "Item",
      },
    ],
    "crafts": 2,
    "disciplines": Array [],
    "id": 5,
    "minRating": null,
    "output": undefined,
    "quantity": 2,
    "recipeId": undefined,
    "type": "Recipe",
  },
  Object {
    "components": Array [
      Object {
        "id": 4,
        "quantity": 1,
        "type": "Item",
      },
    ],
    "crafts": 1,
    "disciplines": Array [],
    "id": 3,
    "minRating": null,
    "output": undefined,
    "quantity": 1,
    "recipeId": undefined,
    "type": "Recipe",
  },
  Object {
    "components": Array [
      Object {
        "id": 3,
        "quantity": 2,
        "type": "Recipe",
      },
      Object {
        "id": 5,
        "quantity": 2,
        "type": "Recipe",
      },
    ],
    "crafts": 2,
    "disciplines": Array [],
    "id": 2,
    "minRating": null,
    "output": undefined,
    "quantity": 2,
    "recipeId": undefined,
    "type": "Recipe",
  },
  Object {
    "components": Array [
      Object {
        "id": 2,
        "quantity": 1,
        "type": "Recipe",
      },
      Object {
        "id": 2,
        "quantity": 1,
        "type": "Recipe",
      },
    ],
    "crafts": 1,
    "disciplines": Array [],
    "id": 1,
    "minRating": null,
    "output": undefined,
    "quantity": 1,
    "recipeId": undefined,
    "type": "Recipe",
  },
]
`;

exports[`craftingSteps gets the correct crafting steps with mystic clovers 1`] = `
Array [
  Object {
    "components": Array [
      Object {
        "id": 99999,
        "quantity": 6,
        "type": "Item",
      },
    ],
    "crafts": 3,
    "disciplines": Array [],
    "id": 19675,
    "minRating": null,
    "output": undefined,
    "quantity": 3,
    "recipeId": undefined,
    "type": "Recipe",
  },
  Object {
    "components": Array [
      Object {
        "id": 4,
        "quantity": 24,
        "type": "Item",
      },
    ],
    "crafts": 12,
    "disciplines": Array [
      "Armorsmith",
    ],
    "id": 7,
    "minRating": 400,
    "output": undefined,
    "quantity": 12,
    "recipeId": 123,
    "type": "Recipe",
  },
  Object {
    "components": Array [
      Object {
        "id": 19675,
        "quantity": 3,
        "type": "Recipe",
      },
    ],
    "crafts": 3,
    "disciplines": Array [],
    "id": 19,
    "minRating": null,
    "output": undefined,
    "quantity": 3,
    "recipeId": undefined,
    "type": "Recipe",
  },
  Object {
    "components": Array [
      Object {
        "id": 19,
        "quantity": 3,
        "type": "Recipe",
      },
    ],
    "crafts": 3,
    "disciplines": Array [],
    "id": 3,
    "minRating": null,
    "output": undefined,
    "quantity": 3,
    "recipeId": undefined,
    "type": "Recipe",
  },
  Object {
    "components": Array [
      Object {
        "id": 7,
        "quantity": 6,
        "type": "Recipe",
      },
    ],
    "crafts": 1,
    "disciplines": Array [],
    "id": 15,
    "minRating": null,
    "output": undefined,
    "quantity": 1,
    "recipeId": undefined,
    "type": "Recipe",
  },
  Object {
    "components": Array [
      Object {
        "id": 7,
        "quantity": 6,
        "type": "Recipe",
      },
      Object {
        "id": 3,
        "quantity": 10,
        "type": "Recipe",
      },
      Object {
        "id": 5,
        "quantity": 4,
        "type": "Recipe",
      },
      Object {
        "id": 15,
        "quantity": 1,
        "type": "Recipe",
      },
    ],
    "crafts": 1,
    "disciplines": Array [],
    "id": 1,
    "minRating": null,
    "output": undefined,
    "quantity": 2,
    "recipeId": undefined,
    "type": "Recipe",
  },
]
`;
